### PR TITLE
slist: abstract node and list implementation

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Singly-linked lists work in a specific way, but this will allow
slist semantics to be defined for any data structure that can store
a "next" pointer in any way at all, possibly something not even
a pointer value, but a table index.

The immediate need for this patch is to allow the easy definition
of an slist variant which stores flags in the low-order bits of the
node pointer.

The compiler does not need them, but the function prototypes have
been left in slist.h to make the header easier to understand by
the end user and not confuse Doxygen.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by lpereira — APPROVED** [review]

**Comment by andrewboie** [comment]
checkpatch failures I believe are all spurious

**Review by andyross — APPROVED** [review]
Other than the inline issue you already caught, this looks great to me.  Very slick.

File: include/misc/slist.h [hunk 1]
```diff
@@ -38,6 +39,7 @@ struct _slist {
 
 typedef struct _slist sys_slist_t;
 
+
```
Comment (jhedberg) [inline include/misc/slist.h]: Nitpick: unrelated change and redundant empty line

I'm curious to see how the new list variants will look like in practice!

**Review by nashif — APPROVED** [review]
